pypeit/core/wavecal/patterns.py

```diff
@@ -0,0 +1,445 @@
+    #pdb.set_trace()
+    dcen = swv_uncertainty*0.8
+    wvcens = np.arange(wvcen-wvoff, wvcen+wvoff+dcen, dcen)
+    # Best
+    if best_dict is None:
+        best_dict = dict(nmatch=0, ibest=-1, bwv=0.)
+
+    # Scan on wv_cen
+    #wvcens = [9400.] # DEBUGGING
+    for ss,iwv_cen in enumerate(wvcens):
+        # Wavelength array
+        wave = iwv_cen + (np.arange(npix) - npix/2.)*disp
+        match_idx, scores = run_quad_match(cut_tcent, wave, wvdata, disp,
+                                           swv_uncertainty=swv_uncertainty,
+                                           pix_tol=pix_tol)
+        # Score
+        mask = np.array([False]*len(cut_tcent))
+        IDs = []
+        for kk,score in enumerate(scores):
+            if score in ['Perf', 'Good', 'Ok']:
+                mask[kk] = True
+                uni, counts = np.unique(match_idx[kk]['matches'], return_counts=True)
+                imx = np.argmax(counts)
+                IDs.append(wvdata[uni[imx]])
+            else:
+                IDs.append(0.)
+        ngd_match = np.sum(mask)
+        #import pdb; pdb.set_trace()
+        # Update in place
+        if ngd_match > best_dict['nmatch']:
+            best_dict['nmatch'] = ngd_match
+            best_dict['midx'] = match_idx
+            best_dict['mask'] = mask
+            best_dict['scores'] = scores
+            best_dict['ibest'] = ss
+            best_dict['bwv'] = iwv_cen
+            best_dict['IDs'] = IDs
+            # Search parameters
+            best_dict['swv_uncertainty'] = swv_uncertainty
+            best_dict['wvoff'] = wvoff
+            best_dict['pix_tol'] = pix_tol
+            best_dict['ampl'] = ampl
+
+
+def score_quad_matches(fidx):
+    """  Grades quad_match results
+    Parameters
+    ----------
+    fidx
+
+    Returns
+    -------
+    scores : list
+
+    """
+    # Loop on indices
+    scores = []
+    for key in fidx.keys():
+        if len(fidx[key]['matches']) == 0:
+            scores.append('None')
+            continue
+        matches = np.array(fidx[key]['matches'])
+        nmatch = matches.size
+        uni, counts = np.unique(matches, return_counts=True)
+        nuni = len(uni)
+        max_counts = max(counts)
+        # Score
+        if (nuni==1) & (nmatch >= 4):
+            scores.append('Perf')
+        elif nmatch == 0:
+            scores.append('None')
+        elif (max_counts == 4) & (nmatch == 5):
+            scores.append('Good')
+        elif (max_counts/nmatch >= 2./3) & (nmatch >= 6):
+            scores.append('Good')
+        elif (nuni == 1) & (nmatch == 3):
+            scores.append('Good')
+        elif (max_counts == 3) & (nmatch == 4):
+            scores.append('OK')
+        elif (nuni == 1) & (nmatch == 2):
+            scores.append('Risk')
+        else:
+            scores.append('Amb')
+    # Return
+    return scores
+
+
+@nb.jit(nopython=True, cache=True)
+def triangles(detlines, linelist, npixels, detsrch=5, lstsrch=10, pixtol=1.0):
```

Comment:
add a brief description.

Reply:
Indeed - I've put a decent description here to explain the pattern matching algorithm.